main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala

```diff
@@ -394,6 +399,26 @@ object Monomorph {
         }
         Expression.Match(visitExp(exp, env0), rs, subst0(tpe), pur, eff, loc)
 
+      case Expression.MatchType(exp, rules, tpe, pur, eff, loc) =>
+        // make the tvars in `exp`'s type rigid
+        // so that Nil: List[x%123] can only match List[_]
+        val renv = exp.tpe.typeVars.foldLeft(RigidityEnv.empty) {
```

Comment:
Why is `subst0` not applied to `exp.tpe` first?

Reply:
What about `Nil: List[a]`? What patterns should match that? Only `List[_]` or more than that?

Reply:
You're right. I need to apply non-strict `subst0` to get the renv.

The important thing that `subst0.nonStrict(exp.tpe)` can contain is free variables, like in your Nil example.

So `Nil: List[a]` where `a` is unbound (i.e. not in the function signature) will _only_ match with `List[_]`